packages/app/src/dimension-mapper/store.tsx

```diff
@@ -0,0 +1,72 @@
+import type { ArrayShape } from '@h5web/shared/hdf5-models';
+import type { PropsWithChildren } from 'react';
+import { createContext, useContext, useEffect, useState } from 'react';
+import type { StoreApi } from 'zustand';
+import { createStore, useStore } from 'zustand';
+
+import { areSameDims } from '../vis-packs/nexus/utils';
+import type { DimensionMapping } from './models';
+
+interface DimMappingState {
+  dims: ArrayShape;
+  axesCount: number;
+  mapping: DimensionMapping;
+  setMapping: (mapping: DimensionMapping) => void;
+  reset: (
+    dims: ArrayShape,
+    axesCount: number,
+    mapping: DimensionMapping,
+  ) => void;
+}
+
+function createLineConfigStore() {
+  return createStore<DimMappingState>((set) => ({
+    dims: [],
+    axesCount: 0,
+    mapping: [],
+    setMapping: (mapping) => set({ mapping }),
+    reset: (dims, axesCount, mapping) => {
+      set({ dims, axesCount, mapping });
+    },
+  }));
+}
+
+const StoreContext = createContext({} as StoreApi<DimMappingState>);
+
+interface Props {}
+export function DimMappingProvider(props: PropsWithChildren<Props>) {
+  const { children } = props;
+
+  const [store] = useState(createLineConfigStore);
+
+  return (
+    <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
+  );
+}
+
+export function useDimMappingState(
+  dims: number[],
+  axesCount: number,
+): [DimensionMapping, (mapping: DimensionMapping) => void] {
+  const state = useStore(useContext(StoreContext));
+
+  /* If current mapping was initialised with different axes count and dimensions,
+   * need to compute new mapping and reset state. */
+  const isStale =
+    axesCount !== state.axesCount || !areSameDims(dims, state.dims);
+
+  const mapping = isStale
+    ? [
+        ...Array.from({ length: dims.length - axesCount }, () => 0),
+        ...(dims.length > 0
+          ? ['y' as const, 'x' as const].slice(-axesCount)
+          : []),
+      ]
+    : state.mapping;
+
+  useEffect(() => {
+    state.reset(dims, axesCount, mapping);
+  }, [isStale]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
Obviously, we don't want to reset the state whenever the `dims` and `mapping` array references change. Hopefully when React 19 comes out and we upgrade to it, we can wrap the `reset` call in an effect event so ESLint doesn't complain about the missing `useEffect` deps.